### ISSUE TITLE
Create DSL for Defining and Laying Out NiftyNodes

### ISSUE THREAD

**Issue by bgroenks96** [nifty-gui/nifty-gui#350]
As of right now, Nifty 2.0 relies solely on Java APIs for designing and maintaining GUI layouts.

There are a few issues with this:
1. Bad for code modularity
   
   GUI code is generally fairly esoteric compared to the rest of an application. As such, it's generally better to keep it as independent as possible (like with MVC model, among other things). Unfortunately, whenever the UI code is written in the same language (and same domain) as the rest of the application code, the two inevitably end up getting intertwined in really nasty ways (see any Swing or SWT application). This is why it's usually beneficial to separate the actual GUI design and layout into a completely separate domain; this is often achieved with XML like in Android development or Nifty 1.x.
2. Not very readable
   
   This is somewhat opinion based, but I think most developers would agree it's kind of hard (or at least harder) to look at general purpose language (Java, C/C++, Python, etc) code and discern what the intended graphical layout actually is. The nice thing about having XML or a DSL in which to define that stuff is that it allows important layout information to be visually accessible; i.e. you don't have to sift through standard language syntax to find the graphics/layout values you're looking for.

That being said, XML sucks. In a lot of ways, it's just as unreadable as Java code for large, complex layout structures. This is in part because it makes you pay a hefty angle-bracket tax.

That's why I'm proposing that we design a DSL to spec and lay out NiftyNodes.

There are a lot of directions we can go with this, but I've provided an example below implementing [this example](https://github.com/void256/nifty-gui/blob/2.0/nifty-examples/src/main/java/de/lessvoid/nifty/examples/usecase/UseCase_a02_QuarterRootNodeWithTwoHorizontalChildNodes.java) in a prototype Nifty DSL.

```
<niftydl version="0.0">
# prototype implementation of example 'UseCase_a02_QuarterRootNodeWithTwoHorizontalChildNodes'

# new root NiftyNode ref 'root'
root:
  layout: vertical
  childLayout: horizontal
  halign: right
  # note: width/height could accept 'px' for pixel width/height
  width: 50%
  height: 50%

  child c1:
    bgcolor: ff08

  child c2:
    bgcolor: ff0f
```

**Comment by relu91** [comment]
I really like the idea to have a DSL for Nifty 2.0. You point out all the major benefit to have it, but it also could make easier not only the layout coding but also some inter operations with application code. This could avoid some major first-user problems like setting up the environment or attach the right listener ( if it will use a swing like pattern for gui events), tedious or duplicate code  and so on. Then, since DSL could be very powerful, I would also give more expressiveness to the user than only layout. Finally I would take a "storage" file format like XML and generate it plus some other java file needed, from the DSL at compile time.  In this way I think that we could go wild in DSL designing and keep a simple storage format, I'd choose json instead XML tough.

**Comment by mkaring** [comment]
If we want to use a precompiled dsl we could compile it straight to java
builder code before the java compiler does it's work. No need for any
storage system this way. It also will not get faster than this any other
way.

Cristiano notifications@github.com schrieb am Di., 18. Aug. 2015 11:08:

**Comment by relu91** [comment]
Yeah I thought it after posting :smile: , but now I'm asking if it could be useful for other use cases. What about modding? A precompiled gui it's not so easy to edit. Or if we what for some reason that the gui is loaded from a Server? I know that we can use a remote class loader but it's a bit tricky. What do you think about these problems?

**Comment by bgroenks96** [comment]
@mkaring Do you really think it would be much more performant to generate builder code instead of interpret the DSL directly at runtime?

Seems like the difference would be pretty marginal; we're talking only minor added overhead from the lazy parsing... do you think that would end up being that significant?

**Comment by mkaring** [comment]
@bgroenks96 Parsing the language at runtime is no big deal at all. I was
more thinking towards the object generation the parser would have to do by
reflection. Reflection is always slow and nasty and I'd like to have ways
to avoid this completely. Also I was thinking towards Nifty on mobile
devices and such where every second during the launch of the application
counts. Again those application don't like Reflection that much, because
they are usually optimized with things like proguard and handling
references by reflection with this is considerably more tricky then static
references.

Static references to Nifty have another advantage. Consider a commercial
product that wants to make some afford to protect the code. Obfuscating
java code is something that can be done with available tools. Obfuscating
some custom DSL is something that requires a lot more work.

I think the nifty way would be to support both. Runtime interpreting of the
language to support desktop applications that are designed to support
modders and compile-time parsing for those who want the UI build (slightly)
faster but get clear static references to Nifty with all the advantages and
added possibilites.

Brian Groenke notifications@github.com schrieb am Do., 20. Aug. 2015 um
08:06 Uhr:

**Comment by bgroenks96** [comment]
@mkaring I like that. Supporting both should be entirely doable.

So now the question is, what do we want the DSL to look like?

**Comment by mkaring** [comment]
@bgroenks96 I think that is something that should be discussed once it is know what nodes are there and can be placed. Because that will be massively different compared to older versions of Nifty.

**Comment by spaceemotion** [comment]
To me the proposed example code looks fairly similar to YAML, which is something that i learned to love and later on learned to hate. You seem to be using spaces and/or tabs in your code, which needs to be consistent across the whole file. Working with multiple people on the file can create problems when some haven't set up their editor(s) correctly, in which case things have to be converted one form or another. Why not add something like the curly brackets lots of languages use?

Just a personal opinion, even though I like the idea of a custom DSL (although just plain CSS would work as well, the only case when coding is required this would require that DSL).

If you want to introduce code segments to those files (even for just simple things like "if clicked, set color to red") the brackets would help enormously as well. From my experience with another project I contributed to, in which they used YAML as a base, not using a custom DSL that supported the braces later on haunted us with some pretty ugly regex parsing to support nested if and if-else statements. Having something like the braces as a start and end indicator would have helped a lot.

Just my two cents :)

**Comment by mkaring** [comment]
The thing is: A compiler/parser/interpreter for any DSL to describe the UI isn't part of the core of Nifty. It is a addition. You will not find a language, everyone likes. I actually like the XML descriptors. I could also work with a json style DSL or even a DSL that looks like java/groovy in the style of gradle scripts.

I think some will like YAML, some a language that looks like python and some a language like CSS (but I think CSS is something that will rather be used for the styling and not for the UI definition).

If we got the facilities to support parsing/compiling a language, I think adding some optional additional languages is actually something that can be done within a very short time. The language itself can't have a fixed definition anyway (it needs to be extendible) to support custom controls and nodes.

**Comment by bgroenks96** [comment]
It sounds like there should be a higher level interface for parsing Nifty UI definitions, with a default implementation of a custom DSL AND pluggable architecture for people to write their own (hopefully with a reasonably low amount of work) parsers for whatever language they might like to use?

**Comment by mkaring** [comment]
@bgroenks96 Something in that direction would be nice I think. Still I think: First things first. Nifty 2.0 needs to be clearly defined so we know what the languages need to be able to do.

**Comment by relu91** [comment]
I agree with @mkaring we need a stable Model of Nifty 2.0 and then write a DSL . In my experience the dsl should be a way to describe user models by using the meta-model which nifty will propose. I agree also that the language should be really flexible to satisfy the needs. I used https://eclipse.org/Xtext/ in my university project. It's really cool and easy to understand but I think it's bound to eclipse IDE. If we use tools like that, making a parser than it's modular it would be a piece of cake.

**Comment by mkaring** [comment]
The general layout of the nodes is taking shape for Nifty 2.0. Also i tricked @void256 into using a very nice system to define the nodes that also allows to discover the nodes that are known to Nifty very easily. We may need to add some additional annotations to the classes to allow proper identification of the properties of each node, but I think that is something we can manage.

As it stands I think we already have a descent base for laying out a basic DSL definition. The design of the node tree will result in a **very** deep tree structure, where each node only contains a few properties.

A simple layout with a coloured bar at the top and bottom already looks like this:

```
RootNode
    DockLayoutNode
        DockLayoutChildNode dock: Top
            FixedSizeLayoutNode height: 30
                BackgroundColorNode color: red
                    ContentNode
        DockLayoutChildNode dock: Bottom
            FixedSizeLayoutNode height: 30
                BackgroundColorNode color: blue
                    ContentNode
```

(this is some language I just made up)

More complex trees will quickly result in very complex trees that will be extremely hard to read. How ever I still strongly believe that we do need at least one DSL definition to support the developers, because as bad as this example up there looks here. Is is far worse when written in Java.

Discovering the nodes that are known to Nifty is pretty easy, because every node has a `NiftyNodeAccessor`. The accessors registered are listed in the meta data of the jar.
For example here: [de.lessvoid.nifty.spi.node.NiftyNodeAccessor](https://github.com/void256/nifty-gui/blob/3ee4bf8a544fdd3b3875ca0f8ac681ee0ba1fbf3/nifty/src/main/resources/META-INF/services/de.lessvoid.nifty.spi.node.NiftyNodeAccessor)

Every provides access to the class of the _adapter_  of the node that provides access to the public methods for the interaction with the node (mostly the getters and setters) and allows fetching the actual implementation of the node from the adapter. Since the adapter already only shows the methods the user is supposed to interact with, this would be the place for any annotation based additional information that helps to define the DSL.

This is basically what is going to happen for the basic nodes. What is currently not defined is how more complex controls will be defined (such as buttons).

**Comment by bgroenks96** [comment]
@void256 @mkaring What's the status of this? Are we ready to start hashing out details regarding a DSL design?

**Comment by mkaring** [comment]
Not sure about that. I mean regarding the structure of the basic nodes it is already pretty clear what will be around. But things like controls, events, controllers of any kind are still totally unclear (to me at least).
